### ISSUE TITLE
dotnet publish copies .cshtml files of dependent projects to the publish directory

### ISSUE THREAD

**Issue by MichaelSpranger** [dotnet/aspnetcore#4330]
### Describe the bug
When running dotnet publish from the command line, under certain conditions the razor (.cshtml) files of precompiled views of dependent projects are copied to the publish directory, in addition to the .Views.dll files containing the precompiled views. I get this consistently in either of two cases:

1. Using the command line option `--no-build` in `dotnet publish`.
2. Setting the property `<GeneratePackageOnBuild>true</GeneratePackageOnBuild>` in the main application's .csproj file.

#### Remarks
- The .cshtml files of the main application do not get copied to the publish directory.
- Setting `<CopyRazorGenerateFilesToPublishDirectory>false</CopyRazorGenerateFilesToPublishDirectory>` makes no difference.

### To Reproduce
1. Create a new ASP.NET core Web application (MVC) with the Visual Studio wizard
1. Add a ASP.NET core Web application (Razor Class Library) project to the solution.
1. On the first project, add a project reference to the second project.
1. Build the solution in Visual Studio
1. Run `dotnet publish WebApplication1/WebApplication1.csproj --no-build' on the command line, where WebApplication1 is the name of the main application.
1. In the folder WebApplication1/bin/Debug/netcoreapp2.1/publish, there is the folder Areas/MyFeature/Pages containing the file Page1.cshtml.

Alternatively, instead of step 5, add `<GeneratePackageOnBuild>true</GeneratePackageOnBuild>` in `WebApplication1.csproj`, and run `dotnet publish WebApplication1/WebApplication1.csproj` without the `--no-build` option instead. 

### Expected behavior
The folder `Areas` and its content should be absent.

### Additional context
```
$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.1.500
 Commit:    b68b931422

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.500\

Host (useful for support):
  Version: 2.1.6
  Commit:  3f4f8eebd8

.NET Core SDKs installed:
  1.0.0-preview2-003131 [C:\Program Files\dotnet\sdk]
  1.0.0-preview2-1-003177 [C:\Program Files\dotnet\sdk]
  1.0.0 [C:\Program Files\dotnet\sdk]
  1.0.1 [C:\Program Files\dotnet\sdk]
  1.0.2 [C:\Program Files\dotnet\sdk]
  1.0.3 [C:\Program Files\dotnet\sdk]
  1.0.4 [C:\Program Files\dotnet\sdk]
  2.0.0 [C:\Program Files\dotnet\sdk]
  2.0.2 [C:\Program Files\dotnet\sdk]
  2.0.3 [C:\Program Files\dotnet\sdk]
  2.1.2 [C:\Program Files\dotnet\sdk]
  2.1.4 [C:\Program Files\dotnet\sdk]
  2.1.100 [C:\Program Files\dotnet\sdk]
  2.1.101 [C:\Program Files\dotnet\sdk]
  2.1.102 [C:\Program Files\dotnet\sdk]
  2.1.103 [C:\Program Files\dotnet\sdk]
  2.1.104 [C:\Program Files\dotnet\sdk]
  2.1.200 [C:\Program Files\dotnet\sdk]
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.403 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
```

**Comment by Eilon** [comment]
@pranavkm / @NTaylorMullen ?

**Comment by NTaylorMullen** [comment]
@pranavkm when was the last time we had a CLI build that has up-to-date SDK bits?

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. We're closing this issue as there was no community involvement here for quite a while now.

**Comment by MichaelSpranger** [comment]
You are not going to fix it?

**Comment by mkArtakMSFT** [comment]
@MichaelSpranger would you be interested in sending us a PR? We'd happily consider it.

**Comment by arex388** [comment]
I also want to join in the discussion that this is a problem, and a very big one for me at least. In my scenario I have two ASP.NET Core 2.2 web apps (A, and B). They share a library project (C) for some common code. In project C I have four Razor views, and I have pre-build scripts in projects A and B to copy those files into them.

All works great in development, but when I publish as Release to IIS. When I do that, the CSHTML files are also published instead of compiled into the *.Views.dll.

I also tried making a Razor class library, and moved views into it. Even though there's a compiled *.Views.dll being published for the RCL, it still publishes the CSHTML files.

At the end of the day, this is a problem for me because those views happen to be my form and table UI templates and they fail to render so in production I just have blank pages.

While I would like to contribute to a fix, I don't have the time nor skills to do so. I'm open to suggestions on how to work around it, and I hope you can fast track this to be fixed in 3.0. @MichaelSimons and I can't possibly be the only ones that have this problem.

The publish needs to compile all views regardless of how they end up in the project or how they're referenced.

I have an open [SO question](https://stackoverflow.com/questions/57584520/some-razor-views-are-not-being-compiled-on-publish-in-visual-studio-2019) on this as well.

Thanks!

**Comment by MichaelSpranger** [comment]
@mkArtakMSFT Unfortunately I do not have enough knowledge on the tooling, so I would not even know where to start looking for the cause(s), let alone fix it. Also, I have since switched jobs and do not work with Asp.Net Core anymore. Trying out a lot of variations to narrow the bug down to the two scenarios and writing the bug report already took me a whole day, for what it's worth.

@arex388 We did our build on Azure DevOps, where I resorted to an additional build task that simply deletes the .cshtml files. Since we had quite many build definitions, this was certainly a nuisance, but served as an acceptable workaround.
That being said, I am not sure if yours is the same problem. With me, the razor views _do_ get included in the appropriate .Views.dll, but in addition to that are _also_ being copied into the publish directory.

**Comment by arex388** [comment]
@MichaelSpranger It's the same issue. Before I wrote my post, I logged into the server and looked through what was published. I found the RCL library and decompiled it with ILSpy and sure enough the views were there *and* they were uploaded as CSHTML files.

Your post gave me an idea though. I logged back into the server and deleted the folder with the CSHTML files and like magic everything started working again. So, thanks for the idea! I publish via FTP so I can't do a "clean up" task after it unfortunately. Oh, well...

**Comment by arex388** [comment]
I take it back, I kind of automated it. I wrote a PowerShell script to check for the offending folder every minute. If it exists, then it is deleted and the apppool is recycled to force rediscovery of the views. I also got rid of the RCL because it literally existed to host four files. I went back to my original pre-build XCOPY and coupled with the PowerShell script I cane make things work. I would like to avoid workarounds like these in the future.
